src/customFileParser/FibonacciDecoder.ts

```diff
@@ -6,50 +6,51 @@ for (let i = 2; i <= FIBONACCI_MAX_LENGTH; i++) {
 }
 
 export default class FibonacciDecoder {
-  public numberRead: number;
+  public numberOfValuesRead = -1;
+  public readonly numberOfValues: number;
   private buffer: ArrayBuffer;
   private data: Uint8Array;
-  private readBitId: number;
-  private currentValue: number;
-  private nextFibIndex: number;
+  private readBitId = -1;
+  private currentValue = -1;
+  private nextFibIndex = -1;
+  private previousBit8 = -1;
 
-  constructor(compressedBuffer: ArrayBuffer, startByte?: number, length?: number) {
-    if (!startByte) { startByte = 0; }
-    if (!length) { length = compressedBuffer.byteLength - startByte; }
-
-    this.numberRead = 0;
+  constructor(compressedBuffer: ArrayBuffer, numberOfValues: number, startByte: number, length: number) {
+    this.numberOfValues = numberOfValues;
     this.buffer = compressedBuffer;
     this.data = new Uint8Array(this.buffer, startByte, length);
+    this.rewind();
+  }
 
-    // read first bit
-    this.currentValue = (this.data[0] & (1 << 7)) >> 7;
-    this.nextFibIndex = 1;
-    this.readBitId = 1;
+  rewind() {
+    this.numberOfValuesRead = 0;
+    this.currentValue = 0;
+    this.previousBit8 = 0;
+    this.nextFibIndex = 0;
+    this.readBitId = 0;
   }
 
-  // Read next encoded value
   nextValue() {
-    this.numberRead++;
+    if (this.numberOfValuesRead >= this.numberOfValues) { throw Error('Reading past end of FibonacciDecoder'); }
+
+    this.numberOfValuesRead++;
 
-    // read rest of file
     for (let i = 0; i < FIBONACCI_MAX_LENGTH; i++) {
-      const currentBit8 = this.data[Math.floor(this.readBitId / 8)] & 1 << (7 - (this.readBitId % 8));
+      const currentByte = this.data[Math.floor(this.readBitId / 8)];
+      const currentBit8 = currentByte & (1 << (7 - (this.readBitId % 8)));
```

Comment:
Much better :D Could maybe add a small comment explaining what actually happens (with an example of a byte and what we end up with)?

Reply:
Done